source/extensions/pageantjacker/pageantjacker.c

```diff
@@ -0,0 +1,186 @@
+					memset((void *) filemap_pointer, 0, AGENT_MAX);
+					if (querylength)
+						memcpy(filemap_pointer, query, querylength);
+
+					dprintf("[PJ(send_query_to_pageant)] Request length: %d. Query buffer preview: %02X:%02X:%02X:%02X:%02X:%02X:%02X:%02X. Request buffer preview: %02X:%02X:%02X:%02X:%02X:%02X:%02X:%02X", querylength, query[0], query[1], query[2], query[3], query[4], query[5], query[6], query[7], filemap_pointer[0], filemap_pointer[1], filemap_pointer[2], filemap_pointer[3], filemap_pointer[4], filemap_pointer[5], filemap_pointer[6], filemap_pointer[7]);
+					
+					// Send the request message to Pageant.
+					dprintf("[PJ(send_query_to_pageant)] Ready to send WM_COPYDATA");
+					if (SendMessage(hPageant, WM_COPYDATA, (WPARAM) NULL, (LPARAM) &pageant_copy_data)) {
+
+						protocol_return_length = get_length_response(filemap_pointer)+4;
+						dprintf("[PJ(send_query_to_pageant)] Result length: %d. Result buffer preview: %02X:%02X:%02X:%02X:%02X:%02X:%02X:%02X", protocol_return_length, filemap_pointer[0], filemap_pointer[1], filemap_pointer[2], filemap_pointer[3], filemap_pointer[4], filemap_pointer[5], filemap_pointer[6], filemap_pointer[7]);
+						if (protocol_return_length && protocol_return_length<AGENT_MAX) {
+							if (ret.blob = calloc(1, protocol_return_length)) {
+								memcpy_result = memcpy(ret.blob, filemap_pointer, protocol_return_length);
+								ret.bloblength = protocol_return_length;
+								ret.result = TRUE;
+								dprintf("[PJ(send_query_to_pageant)] Set Result to TRUE, copied memory to ret.blob (result: 0x%x)",memcpy_result);
+							} else {
+								dprintf("[PJ(send_query_to_pageant)] Malloc error (length: %d).", protocol_return_length);
+								ret.error_message = PAGEANTJACKER_ERROR_ALLOC;
+							}
+						}
+					 } else {
+						// SendMessage failed
+						ret.error_message = PAGEANTJACKER_ERROR_SENDMESSAGE;
+					 }
+					 api_result = UnmapViewOfFile(filemap_pointer);
+					 dprintf("[PJ(send_query_to_pageant)] UnmapViewOfFile returns %d.", api_result);
+				} else {
+					// MapViewOfFile failed
+					ret.error_message = PAGEANTJACKER_ERROR_MAPVIEWOFFILE;
+				}
+				api_result = CloseHandle(filemap);
+				dprintf("[PJ(send_query_to_pageant)] CloseHandle (from CreateFileMapping) returns %d.", api_result);
+			} else {
+				// CreateFileMapping failed
+				ret.error_message = PAGEANTJACKER_ERROR_CREATEFILEMAPPING;
+			}
+		} else {
+			// _snprintf_s failed. Note that this should never happen because it could
+			// mean that somehow %08x has lost its meaning. Essentially though this is
+			// here to guard against buffer overflows.
+			ret.error_message = PAGEANTJACKER_ERROR_REQSTRINGBUILD;
+		}
+
+	} else {
+		// Could not get a handle to Pageant. This probably means that it is not running.
+		ret.error_message = PAGEANTJACKER_ERROR_NOTFOUND;
+	}
+	return ret;
+}
+
+DWORD get_length_response(byte *b) {
+	return (b[3]) | (b[2] << 8) | (b[1] << 16) | (b[0] << 24);
+}
```

Comment:
Yeah that sounds reasonable. The code above, which casts `b` to `DWORD*`, should be fine. You will probably get away with using `DWORD` as the return type as well if you wanted. I'd recommend sticking to the intrinsic though, faster :)

``` c
DWORD get_length_response(PBYTE b)
{
    return (DWORD)__byteswap_ulong(*(DWORD*)b);
}
```
This behaves as expected in my testing, but please validate it yourself too on x64.